pkg/cmd/roachtest/tests/sstable_corruption.go

```diff
@@ -100,13 +100,36 @@ func runSSTableCorruption(ctx context.Context, t test.Test, c cluster.Cluster) {
 		m := c.NewMonitor(ctx)
 		// Run a workload to try to get the node to notice corruption and crash.
 		m.Go(func(ctx context.Context) error {
-			_ = c.RunE(ctx, workloadNode,
-				fmt.Sprintf("./cockroach workload run tpcc --warehouses=100 --tolerate-errors --duration=%s", 2*time.Minute))
+			_ = c.RunE(
+				ctx, workloadNode,
+				"./cockroach workload run tpcc --warehouses=100 --tolerate-errors",
+			)
 			// Don't return an error from the workload. We want outcome of WaitE to be
 			// determined by the monitor noticing that a node died. The workload may
 			// also fail, despite --tolerate-errors, if a node crashes too early.
 			return nil
 		})
+
+		// There's a subtle race condition here that we want to guard against. The
+		// monitor's WaitE waits on either a) the first error from a node that it is
+		// monitoring, OR b) all goroutines derived from the monitor to exit. If the
+		// previous goroutine exits, the wait group may return (with nil) before the
+		// monitor has detected failure from a node. To guard against this, we spawn
+		// a second goroutine on the monitor that acts as a timeout. We expect to
+		// see corruption within the timeout, if not, this goroutine will fail the
+		// test.
+		const timeout = 5 * time.Minute
+		m.Go(func(ctx context.Context) error {
```

Comment:
The "race" you're describing is just that two minutes might not be enough to hit the corruption, right? I'm a little confused by the conjunction of these changes. If the workload now runs "forever" until node failure (since you removed `--duration` from `tpcc`), why do you need an extra timer below?

I think what could still happen is that the workload still returns an error quickly because even with `--tolerate-errors` there might be an early code path or two where an error is returned (during preparing statements for example etc).

The way I would address this is to stick to just one goroutine (the workload one above) but add (this is all inside the `m.Go`)

```go
const maxWait = 10*time.Minute

ctx, cancel := context.WithTimeout(maxWait)
defer cancel()
for {
  err := errors.CombineErrors(
    c.RunE("..."), // NB: != nil since we didn't specify a duration
    errors.New("workload unexpected returned nil"),
  )
  select {
    case <-ctx.Done():
      return ctx.Err()
    case <-time.After(time.Second)
      // loop around
  }
}
```

and then

```go
err := m.WaitE()
if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
  t.Fatalf("no node crash within %s", maxWait)
}
```

The idea is that the will run the workload (perhaps with restarts) for up to 10 minutes. What we expect is that the monitor at some point will tear down because of a node crash, so we'll get a "node x died" error from `WaitE`. (It would be better if `monitor` gave us a structured error; I'll look into that since I'm touching roachtest anyway).
If things go wrong, either the test times out (main ctx cancels) or the ten minutes in the for loop elapse (inner ctx cancels). Either way, `WaitE` gets a ctx cancel error.